Sync/Synchronizers/TabsSynchronizer.swift

```diff
@@ -20,6 +20,78 @@ public class TabsSynchronizer: BaseSingleCollectionSynchronizer, Synchronizer {
         return TabsStorageVersion
     }
 
+    var tabsRecordLastUpload: Timestamp {
+        set(value) {
+            self.prefs.setLong(value, forKey: "lastTabsUpload")
+        }
+
+        get {
+            return self.prefs.unsignedLongForKey("lastTabsUpload") ?? 0
+        }
+    }
+
+    private func createOwnTabsRecord(tabs: [RemoteTab]) -> Record<TabsPayload> {
+        let guid = self.scratchpad.clientGUID
+
+        let jsonTabs: [JSON] = optFilter(tabs.map { $0.toJSON() })
+        let tabsJSON = JSON([
+            "id": guid,
+            "clientName": self.scratchpad.clientName,
+            "tabs": jsonTabs
+        ])
+        log.debug("Sending tabs JSON \(tabsJSON.toString(pretty: true))")
+        let payload = TabsPayload(tabsJSON)
+        return Record(id: guid, payload: payload, ttl: ThreeWeeksInSeconds)
+    }
+
+    // annoyingly can't use direct comparison (==) as our tabs come from the DB with a null client ID as opposed to the remote server who sends us our remote client ID
+    // TODO: there must be a better way to do this but brain freeze right now
+    private func tabRecordsAreEqual(localTabs: [RemoteTab], remoteTabs: [RemoteTab]) -> Bool {
+        if localTabs.count != remoteTabs.count {
+            return false
+        }
+
+        func localTab(localTab: RemoteTab, isSameAsRemoteTab remoteTab: RemoteTab) -> Bool {
+            return remoteTab.URL == localTab.URL && remoteTab.lastUsed == localTab.lastUsed && remoteTab.title == localTab.title && remoteTab.history == localTab.history
+        }
+
+        for (index, tab) in enumerate(localTabs) {
+            // break from the loop the instant we find something has changed
+            if localTab(tab, isSameAsRemoteTab: remoteTabs[index]) {
+                return false
+            }
+        }
+
+        return true
+    }
+
+    private func uploadOurTabs(localTabs: RemoteClientsAndTabs, toServer tabsClient: Sync15CollectionClient<TabsPayload>) -> Success{
+        // check to see if our tabs have changed or we're in a fresh start
+        return tabsClient.get(self.scratchpad.clientGUID) >>== { response in
+            let record = response.value
+            log.debug(record.payload.toString(pretty: true))
+            // get tabs for our local client (no GUID in DB)
+            return localTabs.getTabsForClientWithGUID(nil) >>== { tabs in
+                // if our local version is the same as the one on the remote server, just don't bother uploading again
+                if self.tabRecordsAreEqual(tabs, remoteTabs: record.payload.remoteTabs) {
```

Comment:
Assuming I'm reading this correctly: you're figuring out whether to create and upload a record by… downloading the one on the server, creating the local one, and checking if they're the same.

Don't do that. Let's track (a) when we last uploaded (don't do it if it's too recent), and (b) when we write to the tabs DB, so we know if there are any changes to upload.